File: src/gaet/snapshots.py
"""List local backup snapshots (git tag / git stash list parity)."""

import argparse
import json
from datetime import datetime
from typing import Dict, Any, List

from .core import (
    BACKUP_DIR,
    load_env,
    get_env_int,
    echo,
    box_title,
    box_section,
    status_arrow,
    status_info,
    set_output_modes,
    C, G, Y, D, B, NC,
    DEF_RETENTION_DAYS,
)
from .registry import command


def cmd_snapshots(args: argparse.Namespace) -> None:
    """List local backup snapshots in ~/.gaet/backups/."""
    want_json = getattr(args, "json", False)

    if want_json:
        set_output_modes(quiet=True, plain=True)

    env = load_env()
    retention = get_env_int(env, "GAET_RETENTION_DAYS", DEF_RETENTION_DAYS)

    dump_files = sorted(BACKUP_DIR.glob("*.dump"), key=lambda f: f.stat().st_mtime, reverse=True)

    snapshots_data: List[Dict[str, Any]] = []
    total_bytes = 0

    for f in dump_files:
        st = f.stat()
        size_mb = st.st_size / (1024 * 1024)
        total_bytes += st.st_size
        mtime = datetime.fromtimestamp(st.st_mtime)

        snapshots_data.append({
            "filename": f.name,
            "path": str(f),
            "size_mb": round(size_mb, 1),
            "created_at": mtime.strftime("%Y-%m-%d %H:%M:%S"),
        })

    total_mb = total_bytes / (1024 * 1024)

    if want_json:
        result = {
            "command": "snapshots",
            "count": len(snapshots_data),
            "total_size_mb": round(total_mb, 1),
            "retention_days": retention,
            "snapshots": snapshots_data,
        }
        print(json.dumps(result, indent=2))
        return

    box_title("gaet snapshots")

    if not snapshots_data:
        echo(f"  {Y}No local backup snapshots found in {BACKUP_DIR}.{NC}")
        echo(f"  Run: {C}gaet push{NC} to create your first snapshot.")
        echo()
        return

    box_section(f"Local Snapshots ({len(snapshots_data)} files, {total_mb:.1f} MB total)")
    echo(f"  {B}{'No':<4} {'Snapshot File':<32} {'Size':<10} {'Created At':<20}{NC}")
    echo(f"  {D}{'─'*4} {'─'*32} {'─'*10} {'─'*20}{NC}")

    for idx, snap in enumerate(snapshots_data, 1):
        latest_tag = f" {G}(latest){NC}" if idx == 1 else ""
        echo(f"  {C}[{idx}]{NC:<4} {snap['filename']:<32} {snap['size_mb']:<4.1f} MB   {snap['created_at']:<20}{latest_tag}")

    echo()
    status_info(f"Auto-retention: {retention} days")
    status_info(f"Run: {C}gaet restore <filename.dump>{NC} to restore a snapshot")
    echo()


def _build_snapshots_parser(subparsers, common):
    p = subparsers.add_parser("snapshots", help="List local backup snapshots", parents=[common])
    p.add_argument("--json", action="store_true", help="Output JSON result")
    return p


command("snapshots", "List local backup snapshots", build=_build_snapshots_parser)(cmd_snapshots)

File: pkg/snapshots/snapshots.go
// Package snapshots implements `gaet snapshots`.
package snapshots

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"

	"github.com/ghanirahmans/gaet/pkg/core"
)

// SnapshotInfo holds metadata about a single dump file.
type SnapshotInfo struct {
	Filename  string  `json:"filename"`
	Path      string  `json:"path"`
	SizeMB    float64 `json:"size_mb"`
	CreatedAt string  `json:"created_at"`
	IsLatest  bool    `json:"is_latest,omitempty"`
}

// RunSnapshots implements `gaet snapshots`.
func RunSnapshots(jsonOut bool) error {
	env, _ := core.LoadEnv(core.EnvFile())
	retention := core.GetEnvInt(env, "GAET_RETENTION_DAYS", core.DefRetentionDays)
	backupDir := core.BackupDir()

	matches, _ := filepath.Glob(filepath.Join(backupDir, "*.dump"))
	var infos []SnapshotInfo
	totalBytes := int64(0)

	type entry struct {
		path    string
		modTime time.Time
	}
	var entries []entry
	for _, m := range matches {
		fi, err := os.Stat(m)
		if err == nil {
			entries = append(entries, entry{m, fi.ModTime()})
		}
	}
	sort.Slice(entries, func(i, j int) bool {
		return entries[i].modTime.After(entries[j].modTime)
	})

	for i, e := range entries {
		fi, _ := os.Stat(e.path)
		totalBytes += fi.Size()
		infos = append(infos, SnapshotInfo{
			Filename:  filepath.Base(e.path),
			Path:      e.path,
			SizeMB:    float64(fi.Size()) / 1024 / 1024,
			CreatedAt: e.modTime.Format("2006-01-02 15:04:05"),
			IsLatest:  i == 0,
		})
	}
	totalMB := float64(totalBytes) / 1024 / 1024

	if jsonOut {
		result := map[string]any{
			"command":       "snapshots",
			"count":         len(infos),
			"total_size_mb": fmt.Sprintf("%.1f", totalMB),
			"retention_days": retention,
			"snapshots":     infos,
		}
		enc := json.NewEncoder(os.Stdout)
		enc.SetIndent("", "  ")
		return enc.Encode(result)
	}

	core.BoxTitle("gaet snapshots")
	if len(infos) == 0 {
		core.StatusWarn(fmt.Sprintf("No local backup snapshots found in %s.", backupDir))
		core.Echo(fmt.Sprintf("  Run: %sgaet push%s to create your first snapshot.", core.ColorCyan, core.ColorReset))
		fmt.Println()
		return nil
	}

	core.BoxSection(fmt.Sprintf("Local Snapshots (%d files, %.1f MB total)", len(infos), totalMB))
	fmt.Printf("  %s%-4s %-32s %-10s %-20s%s\n",
		core.ColorBold, "No", "Snapshot File", "Size", "Created At", core.ColorReset)
	fmt.Printf("  %s%s%s\n", core.ColorDim, repeatChar('─', 68), core.ColorReset)

	for i, snap := range infos {
		latest := ""
		if i == 0 {
			latest = fmt.Sprintf(" %s(latest)%s", core.ColorBGreen, core.ColorReset)
		}
		fmt.Printf("  %s[%d]%s  %-32s %5.1f MB  %-20s%s\n",
			core.ColorCyan, i+1, core.ColorReset,
			truncate(snap.Filename, 32),
			snap.SizeMB,
			snap.CreatedAt,
			latest)
	}
	fmt.Println()
	core.StatusInfo(fmt.Sprintf("Auto-retention: %d days", retention))
	core.StatusInfo(fmt.Sprintf("Run: gaet restore <filename.dump> to restore a snapshot"))
	core.PrintDocsFooter()
	return nil
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n-3] + "..."
}

func repeatChar(c rune, n int) string {
	r := make([]rune, n)
	for i := range r {
		r[i] = c
	}
	return string(r)
}
